src/server/pps/server/api_server.go

```diff
@@ -2205,21 +2205,59 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *pps.CreatePipel
 	return &emptypb.Empty{}, nil
 }
 
-func (a *apiServer) createPipeline(ctx context.Context, req *pps.CreatePipelineV2Request) (wrappedJSON string, err error) {
+type defaultsGetter interface {
+	getClusterDefaults(context.Context) (string, error)
+	getProjectDefaults(context.Context, *pfs.Project) (string, error)
+}
+
+func (a *apiServer) makeEffectiveSpec(ctx context.Context, dg defaultsGetter, userSpecJSON string) (effectiveSpecJSON string, effectiveSpec *pps.CreatePipelineRequest, err error) {
+	clusterDefaultsJSON, err := dg.getClusterDefaults(ctx)
+	if err != nil {
+		return "", nil, errors.Wrap(err, "could not get cluster defaults")
+	}
+
+	var req pps.CreatePipelineRequest
+	if err := protojson.Unmarshal([]byte(userSpecJSON), &req); err != nil {
+		return "", nil, errors.Wrapf(err, "could not parse %s as Create Pipeline Request message", userSpecJSON)
+	}
+	projectDefaultsJSON, err := dg.getProjectDefaults(ctx, req.Pipeline.GetProject())
```

Comment:
should project defaults override cluster defaults? if so shouldn't they be used first?